diff --git a/include/osg/GLExtensions b/include/osg/GLExtensions
index 8f047c4ab..654dddc88 100644
--- a/include/osg/GLExtensions
+++ b/include/osg/GLExtensions
@@ -194,6 +194,11 @@ class OSG_EXPORT GLExtensions : public osg::Referenced
         float glVersion;
         float glslLanguageVersion;
 
+        bool isClusteredLightingSupported;
+        bool isDepthClampSupported;
+        bool isComputeShaderRipplesSupported;
+        bool isPolygonModeSupported;
+
         bool isGlslSupported;
         bool isShaderObjectsSupported;
         bool isVertexShaderSupported;
diff --git a/src/osg/GLExtensions.cpp b/src/osg/GLExtensions.cpp
index 357796b17..22989f7e0 100644
--- a/src/osg/GLExtensions.cpp
+++ b/src/osg/GLExtensions.cpp
@@ -451,6 +455,34 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
     glVersion = validContext ? findAsciiToFloat( versionString ) : 0.0f;
     glslLanguageVersion = 0.0f;
 
+    // Custered lighting support check
+    isClusteredLightingSupported = false;
+    if (validContext) {
+        GLint maxFragmentSSBOs = 0;
+        GLint maxVertexSSBOs = 0;
+        GLint maxComputeSSBOs = 0;
+
+        glGetIntegerv(GL_MAX_FRAGMENT_SHADER_STORAGE_BLOCKS, &maxFragmentSSBOs);
+        glGetIntegerv(GL_MAX_VERTEX_SHADER_STORAGE_BLOCKS, &maxVertexSSBOs);
+        glGetIntegerv(GL_MAX_COMPUTE_SHADER_STORAGE_BLOCKS, &maxComputeSSBOs);
+
+        if (maxFragmentSSBOs >= 5 && maxVertexSSBOs >= 5 && maxComputeSSBOs >= 5)
+            isClusteredLightingSupported = true;
+    }
+
+    isComputeShaderRipplesSupported = false;
+    if (validContext) {
+        GLint maxComputeWorkGroupsInvocations = 0;
+
+        glGetIntegerv(/*GL_MAX_COMPUTE_WORK_GROUP_INVOCATIONS*/0x90EB, &maxComputeWorkGroupsInvocations);
+
+        if (maxComputeWorkGroupsInvocations >= 256)
+            isComputeShaderRipplesSupported = true;
+    }
+
+    isDepthClampSupported = isGLExtensionSupported(contextID, "GL_EXT_depth_clamp");
+    isPolygonModeSupported = isGLExtensionSupported(contextID, "GL_NV_polygon_mode");
+
     bool shadersBuiltIn = OSG_GLES2_FEATURES || OSG_GLES3_FEATURES || OSG_GL3_FEATURES;
 
     isShaderObjectsSupported = validContext && (shadersBuiltIn || osg::isGLExtensionSupported(contextID,"GL_ARB_shader_objects"));
@@ -479,6 +511,7 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
 
     isClipControlSupported = validContext &&
                              (isGLExtensionSupported(contextID,"GL_ARB_clip_control") ||
+                             isGLExtensionSupported(contextID,"GL_EXT_clip_control") ||
                              (glVersion >= 4.5f));
 
 
@@ -620,7 +653,7 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
     setGLExtensionFuncPtr(glUniformMatrix4x3fv, "glUniformMatrix4x3fv", validContext);
 
     // ARB_clip_control
-    setGLExtensionFuncPtr(glClipControl, "glClipControl", validContext);
+    setGLExtensionFuncPtr(glClipControl, "glClipControl", "glClipControlEXT", validContext);
 
     // EXT_geometry_shader4
     setGLExtensionFuncPtr(glProgramParameteri,  "glProgramParameteri", "glProgramParameteriEXT", validContext);
@@ -850,7 +883,7 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
                                  isGLExtensionOrVersionSupported(contextID,"GL_EXT_multitexture", 1.3f));
 
     isTextureFilterAnisotropicSupported = validContext && isGLExtensionSupported(contextID,"GL_EXT_texture_filter_anisotropic");
-    isTextureSwizzleSupported = validContext && isGLExtensionSupported(contextID,"GL_ARB_texture_swizzle");
+    isTextureSwizzleSupported = validContext && (isGLExtensionSupported(contextID,"GL_ARB_texture_swizzle"));
     isTextureCompressionARBSupported = validContext && (builtInSupport || isGLExtensionOrVersionSupported(contextID,"GL_ARB_texture_compression", 1.3f));
     isTextureCompressionS3TCSupported = validContext && (isGLExtensionSupported(contextID,"GL_EXT_texture_compression_s3tc") ||
                                                          isGLExtensionSupported(contextID, "GL_S3_s3tc") ||
@@ -873,6 +906,10 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
                                                          isGLExtensionSupported(contextID, "GL_OES_texture_compression_astc") ||
                                                          isGLExtensionSupported(contextID, "WEBGL_compressed_texture_astc"));
 
+    // Fake support, decompress in texture.c
+    isTextureCompressionS3TCSupported = true;
+    isTextureCompressionRGTCSupported = true;
+
     isTextureMirroredRepeatSupported = validContext &&
                                        (builtInSupport ||
                                         isGLExtensionOrVersionSupported(contextID,"GL_IBM_texture_mirrored_repeat", 1.4f) ||
@@ -892,7 +930,7 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
     isGenerateMipMapSupported = validContext && (builtInSupport || isGLExtensionOrVersionSupported(contextID,"GL_SGIS_generate_mipmap", 1.4f));
     preferGenerateMipmapSGISForPowerOfTwo = (radeonHardwareDetected||fireGLHardwareDetected) ? false : true;
     isTextureMultisampledSupported = validContext && (isGLExtensionSupported(contextID,"GL_ARB_texture_multisample"));
-    isShadowSupported = validContext && (OSG_GL3_FEATURES || isGLExtensionSupported(contextID,"GL_ARB_shadow"));
+    isShadowSupported = validContext && (OSG_GL3_FEATURES || OSG_GLES3_FEATURES || isGLExtensionSupported(contextID,"GL_ARB_shadow"));
     isShadowAmbientSupported = validContext && (isGLExtensionSupported(contextID,"GL_ARB_shadow_ambient"));
     isClientStorageSupported = validContext && (isGLExtensionSupported(contextID,"GL_APPLE_client_storage"));
     isNonPowerOfTwoTextureMipMappedSupported = validContext && (builtInSupport || isGLExtensionSupported(contextID, "GL_OES_texture_npot") || isGLExtensionOrVersionSupported(contextID, "GL_ARB_texture_non_power_of_two", 2.0));
@@ -919,10 +957,10 @@ GLExtensions::GLExtensions(unsigned int in_contextID):
     isTextureLODBiasSupported  = (glVersion >= 3.0f) || isGLExtensionSupported(contextID, "GL_EXT_texture_lod_bias");
 #else
     isTextureMaxLevelSupported = (glVersion >= 1.2f);
-    isTextureLODBiasSupported  = (glVersion >= 1.2f) || isGLExtensionSupported(contextID, "GL_EXT_texture_lod_bias");
+    isTextureLODBiasSupported  = false;//(glVersion >= 1.2f) || isGLExtensionSupported(contextID, "GL_EXT_texture_lod_bias");
 #endif
 
-    isTextureStorageEnabled = validContext && ((glVersion >= 4.2f) || isGLExtensionSupported(contextID, "GL_ARB_texture_storage"));
+    isTextureStorageEnabled = validContext && (OSG_GL3_FEATURES || OSG_GLES3_FEATURES || (glVersion >= 4.2f) || isGLExtensionSupported(contextID, "GL_ARB_texture_storage"));
 
     if (isTextureStorageEnabled)
     {
diff --git a/src/osg/PolygonMode.cpp b/src/osg/PolygonMode.cpp
index 35db6758f..f6264e696 100644
--- a/src/osg/PolygonMode.cpp
+++ b/src/osg/PolygonMode.cpp
@@ -13,6 +13,8 @@
 #include <osg/GL>
 #include <osg/PolygonMode>
 #include <osg/Notify>
+#include <osg/State>
+#include <osg/GLExtensions>
 
 using namespace osg;
 
@@ -66,12 +68,15 @@ PolygonMode::Mode PolygonMode::getMode(Face face) const
     return _modeFront;
 }
 
-void PolygonMode::apply(State&) const
+void PolygonMode::apply(State& state) const
 {
-#if !defined(OSG_GLES1_AVAILABLE) && !defined(OSG_GLES2_AVAILABLE) && !defined(OSG_GLES3_AVAILABLE)
+    const GLExtensions* extensions = state.get<GLExtensions>();
+    if (!extensions->isPolygonModeSupported)
+        return;
+
     if (_modeFront==_modeBack)
     {
-        glPolygonMode(GL_FRONT_AND_BACK,(GLenum)_modeFront);
+        glPolygonModeNV(GL_FRONT_AND_BACK,(GLenum)_modeFront);
     }
     else
     {
@@ -82,8 +87,5 @@ void PolygonMode::apply(State&) const
             OSG_NOTICE << "Warning: PolygonMode::apply(State&) - only GL_FRONT_AND_BACK is supported." << std::endl;
         #endif
     }
-#else
-    OSG_NOTICE << "Warning: PolygonMode::apply(State&) - is not supported." << std::endl;
-#endif
 }
 
index 152decdcc..778f00645 100644
--- a/src/osg/State.cpp
+++ b/src/osg/State.cpp
@@ -297,6 +297,25 @@ void State::initializeExtensionProcs()
     #endif
     }
 
+    setModeValidity(GL_NORMALIZE,      false); // 0x0BA1
+
+    if (!osg::isGLExtensionSupported(_contextID,"GL_EXT_clip_cull_distance"))
+    {
+        for (int i = 0; i < 8; i++)
+            setModeValidity(0x3000 + i,    false); // GL_CLIP_PLANE0
+        OSG_NOTICE<<"GL_EXT_clip_cull_distance not supported."<<std::endl;
+    }
+    if (!osg::isGLExtensionSupported(_contextID,"GL_EXT_depth_clamp"))
+    {
+        setModeValidity(0x864F,    false); // GL_DEPTH_CLAMP
+        OSG_NOTICE<<"GL_EXT_depth_clamp not supported."<<std::endl;
+    }
+    if (!osg::isGLExtensionSupported(_contextID,"GL_NV_polygon_mode"))
+    {
+        setModeValidity(0x2A01,    false); // GL_POLYGON_OFFSET_POINT
+        setModeValidity(0x2A02,    false); // GL_POLYGON_OFFSET_LINE
+        OSG_NOTICE<<"GL_NV_polygon_mode not supported."<<std::endl;
+    }
 
     _extensionProcsInitialized = true;
 
