diff --git a/apps/openmw/engine.cpp b/apps/openmw/engine.cpp
index 8d8a0bc7b9..abe143fa03 100644
--- a/apps/openmw/engine.cpp
+++ b/apps/openmw/engine.cpp
@@ -487,6 +487,9 @@ void OMW::Engine::setSkipMenu(bool skipMenu, bool newGame)
     mNewGame = newGame;
 }
 
+// To share the viewer with Android interfaces
+osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 void OMW::Engine::createWindow()
 {
     const int screen = Settings::video().mScreen;
@@ -693,6 +696,9 @@ void OMW::Engine::createWindow()
 
     mViewer->getEventQueue()->getCurrentEventState()->setWindowRectangle(
         0, 0, graphicsWindow->getTraits()->width, graphicsWindow->getTraits()->height);
+
+    // To share the viewer with Android interfaces
+    g_viewer = mViewer;
 }
 
 void OMW::Engine::setWindowIcon()
@@ -1070,6 +1076,8 @@ void OMW::Engine::go()
 
     mLuaWorker->join();
 
+    g_viewer.release();
+
     // Save user settings
     Settings::Manager::saveUser(mCfgMgr.getUserConfigPath() / "settings.cfg");
     Settings::ShaderManager::get().save();
diff --git a/files/data/mygui/openmw_postprocessor_hud.layout b/files/data/mygui/openmw_postprocessor_hud.layout
index 4ed8ecdb36..758b1b2e99 100644
--- a/files/data/mygui/openmw_postprocessor_hud.layout
+++ b/files/data/mygui/openmw_postprocessor_hud.layout
@@ -3,7 +3,7 @@
   <Widget type="Window" skin="MW_Window" position="0 0 500 500" layer="Debug" name="_Main" align="Stretch">
     <Property key="Visible" value="false"/>
     <Property key="Caption" value="#{OMWShaders:PostProcessHUD}"/>
-    <Property key="MinSize" value="600 600"/>
+    <Property key="MinSize" value="600 430"/>
 
     <Widget type="TabControl" skin="TabControl_NoBorder" position="8 8 468 450" align="Stretch" name="TabControl">
         <Property key="ButtonAutoWidth" value="true"/>
@@ -20,7 +20,7 @@
             <Property key="Padding" value="0"/>
             <Property key="Spacing" value="8"/>
 
-            <Widget type="Widget" position ="0 0 0 300">
+            <Widget type="Widget" position ="0 0 0 250">
                 <UserString key="HStretch" value="true"/>
                 <UserString key="VStretch" value="false"/>
 
