diff --git a/apps/openmw/mwrender/localmap.cpp b/apps/openmw/mwrender/localmap.cpp
index ade0faafac..278da9c435 100644
--- a/apps/openmw/mwrender/localmap.cpp
+++ b/apps/openmw/mwrender/localmap.cpp
@@ -742,6 +742,8 @@ namespace MWRender
         stateset->addUniform(new osg::Uniform("skyBlendingStart", 8000000.0f));
         stateset->addUniform(new osg::Uniform("screenRes", osg::Vec2f{ 1, 1 }));
 
+        stateset->addUniform(new osg::Uniform("isReflection", false), osg::StateAttribute::ON | osg::StateAttribute::OVERRIDE);
+
         osg::ref_ptr<SceneUtil::Light> light = new SceneUtil::Light;
         light->setPosition(osg::Vec4(-0.3f, -0.3f, 0.7f, 0.f));
         light->setDiffuse(osg::Vec4(0.7f, 0.7f, 0.7f, 1.f));
diff --git a/apps/openmw/mwsound/ffmpegdecoder.cpp b/apps/openmw/mwsound/ffmpegdecoder.cpp
index 406bd47d66..ade613f13b 100644
--- a/apps/openmw/mwsound/ffmpegdecoder.cpp
+++ b/apps/openmw/mwsound/ffmpegdecoder.cpp
@@ -294,7 +294,7 @@ namespace MWSound
         }
 
         // avformat_open_input frees the user supplied AVFormatContext on failure
-        if (avformat_open_input(&ctx, name, fmt, nullptr) != 0)
+        if (avformat_open_input(&ctx, name, const_cast<AVInputFormat*>(fmt), nullptr) != 0)
             return false;
 
         formatCtx.reset(std::exchange(ctx, nullptr));
diff --git a/files/shaders/compatibility/water.frag b/files/shaders/compatibility/water.frag
index e202c3b5bc..6fa32f2758 100644
--- a/files/shaders/compatibility/water.frag
+++ b/files/shaders/compatibility/water.frag
@@ -206,10 +206,10 @@ void main(void)
     rawRefraction = refraction;
 
     // brighten up the refraction underwater
-    if (cameraPos.z < 0)
+    if (cameraPos.z < 0.0)
         refraction = clamp(refraction * 1.5, 0.0, 1.0);
 
-    if (cameraPos.z >= 0)
+    if (cameraPos.z >= 0.0)
     {
         vec3 deepWaterColor = waterColor * fogScale + fogOffset;
         if (!hasOpaqueGeometry(refractionCoords))
diff --git a/files/shaders/core/lighting/cluster.comp b/files/shaders/core/lighting/cluster.comp
index 2ef435b483..ce5ba77470 100644
--- a/files/shaders/core/lighting/cluster.comp
+++ b/files/shaders/core/lighting/cluster.comp
@@ -30,16 +30,16 @@ vec3 screenToView(vec2 screenCoord)
 void main()
 {
     uint tileIndex = gl_WorkGroupID.x + (gl_WorkGroupID.y * uint(gridSize.x)) + (gl_WorkGroupID.z * uint(gridSize.x) * uint(gridSize.y));
-    vec2 tileSize = screenRes / ivec2(gridSize.xy);
+    vec2 tileSize = vec2(ivec2(screenRes) / ivec2(gridSize.xy));
 
-    vec2 minTile_screenspace = gl_WorkGroupID.xy * tileSize;
-    vec2 maxTile_screenspace = (gl_WorkGroupID.xy + 1) * tileSize;
+    vec2 minTile_screenspace = vec2(gl_WorkGroupID.xy) * tileSize;
+    vec2 maxTile_screenspace = vec2(gl_WorkGroupID.xy + 1u) * tileSize;
 
     vec3 minTile = screenToView(minTile_screenspace);
     vec3 maxTile = screenToView(maxTile_screenspace);
 
-    float planeNear = near * pow(clusterFar / near, gl_WorkGroupID.z / gridSize.z);
-    float planeFar = near * pow(clusterFar / near, (gl_WorkGroupID.z + 1) / gridSize.z);
+    float planeNear = near * pow(clusterFar / near, float(gl_WorkGroupID.z) / gridSize.z);
+    float planeFar = near * pow(clusterFar / near, float(gl_WorkGroupID.z + 1u) / gridSize.z);
 
     vec3 minPointNear = lineIntersectionWithZPlane(vec3(0, 0, 0), minTile, planeNear);
     vec3 minPointFar = lineIntersectionWithZPlane(vec3(0, 0, 0), minTile, planeFar);
diff --git a/files/shaders/core/lighting/cull.comp b/files/shaders/core/lighting/cull.comp
index 43772b2df1..379ddbffa5 100644
--- a/files/shaders/core/lighting/cull.comp
+++ b/files/shaders/core/lighting/cull.comp
@@ -29,16 +29,16 @@ bool testSphereAABB(int i, Cluster cluster)
 
 void main()
 {
-    if (gl_LocalInvocationIndex == 0) {
-        g_Count = 0;
+    if (gl_LocalInvocationIndex == 0u) {
+        g_Count = 0u;
         g_cluster = clusters[gl_WorkGroupID.x];
     }
 
     barrier();
 
-    uint index = 0;
+    uint index = 0u;
 
-    for (uint i = gl_LocalInvocationIndex; i < pointLight.length(); i += @workGroupSize) {
+    for (uint i = gl_LocalInvocationIndex; i < uint(pointLight.length()); i += uint(@workGroupSize)) {
         PointLight light = pointLight[i];
 
         vec3 center = light.position.xyz;
@@ -48,9 +48,9 @@ void main()
         vec3 aabbMax = g_cluster.maxPoint.xyz;
 
         if (sphereAABBIntersection(center, radius, aabbMin, aabbMax)) {
-            index = atomicAdd(g_Count, 1);
+            index = atomicAdd(g_Count, 1u);
 
-            if (index >= @maxLightsPerCluster) break;
+            if (index >= uint(@maxLightsPerCluster)) break;
 
             g_List[index] = i;
         }
@@ -58,14 +58,14 @@ void main()
 
     barrier();
 
-    if (gl_LocalInvocationIndex == 0) {
+    if (gl_LocalInvocationIndex == 0u) {
         g_Offset = atomicAdd(globalLightIndexCount, g_Count);
         lightGrid[gl_WorkGroupID.x] = LightGrid(g_Offset, g_Count);
     }
 
     barrier();
 
-    for (uint i = gl_LocalInvocationIndex; i < g_Count; i += @workGroupSize) {
+    for (uint i = gl_LocalInvocationIndex; i < g_Count; i += uint(@workGroupSize)) {
         lightIndexList[g_Offset + i] = g_List[i];
     }
 }
diff --git a/files/shaders/core/ripples_blobber.comp b/files/shaders/core/ripples_blobber.comp
index 235f58a5ba..81e849b2f4 100644
--- a/files/shaders/core/ripples_blobber.comp
+++ b/files/shaders/core/ripples_blobber.comp
@@ -16,7 +16,7 @@ uniform vec2 offset;
 
 void main()
 {
-    ivec2 texel = ivec2(gl_GlobalInvocationID.xy + offset);
+    ivec2 texel = ivec2(gl_GlobalInvocationID.xy + uvec2(offset));
 
     vec4 color = imageLoad(imageIn, texel);
     float wavesizeMultiplier = getTemporalWaveSizeMultiplier(osg_SimulationTime);
